state/txindex/kv/kv.go

```diff
@@ -613,46 +609,83 @@ LOOP:
+//       Usually something like "tx.height" or "account.owner"
+//    2. A value. That corresponds to the key. In the above
+//       example the value could be "5" or "Ivan"
+//    3. The height of the Tx that aligns with the key and value.
+//    4. The index of the Tx that aligns with the key and value
+
+// The event key
+func key(compositeKey, value string, height int64, index uint32) []byte {
+	key, err := orderedcode.Append(
+		nil,
+		compositeKey,
+		value,
+		height,
+		int64(index),
+	)
+	if err != nil {
+		panic(err)
+	}
+	return key
+}
 
-func isTagKey(key []byte) bool {
-	return strings.Count(string(key), tagKeySeparator) == 3
+// parseValueFromKey parses an event key and extracts out the value, returning an error if one arises.
+// This will also involve ensuring that the key has the correct format.
+// CONTRACT: function doesn't check that the prefix is correct. This should have already been done by the iterator
+func parseValueFromKey(key []byte) (string, error) {
+	var (
+		compositeKey, value string
+		height, index       int64
+	)
+	remaining, err := orderedcode.Parse(string(key), &compositeKey, &value, &height, &index)
+	if err != nil {
+		return "", err
+	}
+	if len(remaining) != 0 {
+		return "", fmt.Errorf("unexpected remainder in key: %s", remaining)
+	}
+	return value, nil
 }
 
-func extractValueFromKey(key []byte) string {
-	parts := strings.SplitN(string(key), tagKeySeparator, 3)
-	return parts[1]
+func keyFromEvent(compositeKey string, value []byte, result *abci.TxResult) []byte {
+	return key(compositeKey, string(value), result.Height, result.Index)
 }
 
-func keyForEvent(key string, value []byte, result *abci.TxResult) []byte {
-	return []byte(fmt.Sprintf("%s/%s/%d/%d",
-		key,
-		value,
-		result.Height,
-		result.Index,
-	))
+func keyFromHeight(result *abci.TxResult) []byte {
+	return key(types.TxHeightKey, fmt.Sprintf("%d", result.Height), result.Height, result.Index)
 }
 
-func keyForHeight(result *abci.TxResult) []byte {
-	return []byte(fmt.Sprintf("%s/%d/%d/%d",
-		types.TxHeightKey,
-		result.Height,
-		result.Height,
-		result.Index,
-	))
+// Prefixes: these represent an initial part of the key and are used by iterators to iterate over a small
+// section of the kv store during searches.
+
+func prefixFromCompositeKey(compositeKey string) []byte {
+	key, _ := orderedcode.Append(nil, compositeKey)
+	return key
 }
 
-func startKeyForCondition(c query.Condition, height int64) []byte {
-	if height > 0 {
-		return startKey(c.CompositeKey, c.Operand, height)
-	}
-	return startKey(c.CompositeKey, c.Operand)
+func prefixFromCompositeKeyAndValue(compositeKey, value string) []byte {
+	key, _ := orderedcode.Append(nil, compositeKey, value)
+	return key
 }
 
-func startKey(fields ...interface{}) []byte {
-	var b bytes.Buffer
-	for _, f := range fields {
-		b.Write([]byte(fmt.Sprintf("%v", f) + tagKeySeparator))
+func prefixFromCompositeKeyValueAndHeight(compositeKey, value string, height int64) []byte {
```

Comment:
There might be something a little more pretty than what I've done here that someone can suggest. In Javascript you have optional arguments. But if anything I wanted to be explicit with the structure of the keys

Reply:
Would variadic arguments help? https://gobyexample.com/variadic-functions